### ISSUE TITLE
bind: cannot assign requested address

### ISSUE THREAD

**Issue by ccRidge** [AdguardTeam/AdGuardHome#5790]
### Prerequisites

- [X] I have checked the [Wiki](https://github.com/AdguardTeam/AdGuardHome/wiki) and [Discussions](https://github.com/AdguardTeam/AdGuardHome/discussions) and found no answer

- [X] I have searched other issues and found no duplicates

- [X] I want to report a bug and not ask a question

### Operating system type

Linux, OpenWrt

### CPU architecture

64-bit ARM

### Installation

Other (please mention in the description)

### Setup

On a router, DHCP is handled by the router

### AdGuard Home version

v0.107.21

#### Installation
OpenWRT opkg

#### What did you do?
Reboot the router.

#### Expected result
AdGuard Home should start reliably after a system reboot.

#### Actual result
Fairly often, AdGuard Home errors out on reboot.  Failed 3 out of 4 times while trying to gather a log file that demonstrates success for this ticket.  Failure message is:

> [fatal] couldn't start forwarding DNS server: listening to udp socket: listen udp 127.0.0.1:53: bind: cannot assign requested address

#### Workaround
SSH into the router and restart AdGuard Home through init.d script.  This method is 100% successful.
Alternatively, reboot again, which is sometimes successful.    

#### Additional information
I've provided three log files for comparison.
1. reboot-success.log - verbose logs following a reboot, where AdGuard Home starts up successfully.
2. reboot-fail.log  - verbose logs following a reboot, where AdGuard Home fails to start up.
3. init_d-success.log - verbose logs following a manual restart using the init.d script, where AdGuard Home starts up successfully. 

[reboot-success.log](https://github.com/AdguardTeam/AdGuardHome/files/11362133/reboot-success.log)
[reboot-fail.log](https://github.com/AdguardTeam/AdGuardHome/files/11362134/reboot-fail.log)
[init_d-success.log](https://github.com/AdguardTeam/AdGuardHome/files/11362135/init_d-success.log)

#### Observations
I noticed that for the init.d restart instances, there is a 10-second delay between log entries for "Go to http" style messages, and the "Creating the UDP server socket" messages.  For instance:

  > 2023/04/30 18:36:30.996889 3586#1 [info] Go to http://[fe80::ea9f:80ff:fef0:6a84%wlan1]:8080
  > 2023/04/30 18:36:41.302740 3586#34 [info] Creating the UDP server socket

This is interesting, but probably a red herring since there is no delay in either of the failed or success logs for the reboot scenario.

#### Things I've tried
* I tried delaying the start of AdGuard Home by setting "START=99" in the init.d script.  This did not help.
* My router is using DNSMASQ as an upstream DNS, with port configured to 54, as described here: [https://forum.openwrt.org/t/how-to-updated-2021-installing-adguardhome-on-openwrt-manual-and-opkg-method/113904.](https://forum.openwrt.org/t/how-to-updated-2021-installing-adguardhome-on-openwrt-manual-and-opkg-method/113904)  I thought maybe there could be a race condition between DNSMASQ coming up defaulted to port 53 after a reboot, but not seeing anything in logs to corroborate that 
* The WAN network of my router is configured to use unencrypted DNS to work around NTP issues as described here: [https://openwrt.org/docs/guide-user/services/dns/adguard-home#bypassing_encrypted_dns_for_ntp.](https://openwrt.org/docs/guide-user/services/dns/adguard-home#bypassing_encrypted_dns_for_ntp)

#### Mea Culpa
There is at least some evidence that this issue has been reported, for instance here: [https://github.com/AdguardTeam/AdGuardHome/issues/3760.](https://github.com/AdguardTeam/AdGuardHome/issues/3760)  Although I haven't found anything active for OpenWRT, hence the new ticket.

**Comment by gustakasn0v** [comment]
Got a similar issue, except in my case it fails to bind to the web UI (tcp) port, not the (udp) DNS port. In my case a restart fixes it as well:

```
2024/01/23 21:42:19.392527 [info] AdGuard Home, version v0.107.36
2024/01/23 21:42:19.418316 [info] tls: using default ciphers
2024/01/23 21:42:19.514381 [info] safesearch default: disabled
2024/01/23 21:42:19.552858 [info] Initializing auth module: /mnt/sda1/adguardhome/data/sessions.db
2024/01/23 21:42:19.563335 [info] auth: initialized.  users:1  sessions:2
2024/01/23 21:42:19.563810 [info] AdGuard Home updates are disabled
2024/01/23 21:42:19.566420 [info] web: initializing
2024/01/23 21:42:20.135131 [info] dnsproxy: cache: enabled, size 4096 b
2024/01/23 21:42:20.135174 [info] dnsproxy: max goroutines is set to 300
2024/01/23 21:42:20.143316 [info] clients: processing addresses
2024/01/23 21:42:20.144750 [info] AdGuard Home is available at the following addresses:
2024/01/23 21:42:20.144788 [info] go to http://192.168.4.2:8080
2024/01/23 21:42:20.162190 [info] stopped
2024/01/23 21:42:20.162263 [fatal] listen tcp 192.168.4.2:8080: bind: cannot assign requested address
```

I was able to permanently fix this by replacing the bind addresses (`http.address` and `dns.bind_hosts` in `/etc/adguardhome.yaml`) with `0.0.0.0` instead of the LAN IP address of my device; this is *relatively* safe for me because my device is not a router, but I wouldn't advise it for gateway routers as 0.0.0.0 would mean binding to the WAN port which can be unsafe.

This does narrow the issue down to some race condition between OpenWRT configuring the network interfaces and AdGuard Home starting. Notably, `/etc/init.d/adguardhome` says:

```
...
# starts just after network starts to avoid some network race conditions
START=25
...
```
Whereas `/etc/init.d/network` has a start priority of `20`, so AdGuardHome *should* be waiting to networking to come up. Not sure if these start scripts run in parallel however, meaning there's a chance AdGuardHome tries to start before the network *finishes* coming up

**Comment by overwatch3560** [comment]
@ccRidge is this still an issue on the most recent version?

**Comment by Aejs** [comment]
@overwatch3560 im Running on v0.107.47 and got the same Bug.

2024-04-04T20:24:49.788044364Z 2024/04/04 22:24:49.787998 [info] AdGuard Home updates are disabled
2024-04-04T20:24:49.788047100Z 2024/04/04 22:24:49.788001 [info] web: initializing
2024-04-04T20:24:49.913329676Z 2024/04/04 22:24:49.913137 [info] dnsproxy: cache: enabled, size 4096 b
2024-04-04T20:24:49.913372457Z 2024/04/04 22:24:49.913167 [info] dnsproxy: max goroutines is set to 300
2024-04-04T20:24:49.913457799Z 2024/04/04 22:24:49.913371 [info] dnsproxy: cache: disabled
2024-04-04T20:24:49.913497926Z 2024/04/04 22:24:49.913387 [info] Ratelimit is enabled and set to 200 rps, IPv4 subnet mask len 24, IPv6 subnet mask len 56
2024-04-04T20:24:49.913502153Z 2024/04/04 22:24:49.913389 [info] dnsproxy: server will refuse requests of type ANY
2024-04-04T20:24:49.913505099Z 2024/04/04 22:24:49.913391 [info] dnsproxy: cache: enabled, size 536870912 b
2024-04-04T20:24:49.913507524Z 2024/04/04 22:24:49.913393 [info] dnsproxy: max goroutines is set to 300
2024-04-04T20:24:49.913658461Z 2024/04/04 22:24:49.913587 [info] AdGuard Home is available at the following addresses:
2024-04-04T20:24:49.913671837Z 2024/04/04 22:24:49.913600 [info] go to http://172.21.0.16:80
2024-04-04T20:24:49.913891243Z 2024/04/04 22:24:49.913752 [info] clients: processing addresses
2024-04-04T20:24:49.913961166Z 2024/04/04 22:24:49.913901 [info] stopped
2024-04-04T20:24:49.913971396Z 2024/04/04 22:24:49.913921 [fatal] listen tcp 172.21.0.16:80: bind: cannot assign requested address

**Comment by ingstr-com** [comment]
I just had the same problem as @overwatch3560.
I enabled the verbose logs but no logs are saved to the file, only the same output in the docker logs as @overwatch3560 above.

**Comment by ericegonz** [comment]
> I was able to permanently fix this by replacing the bind addresses (`http.address` and `dns.bind_hosts` in `/etc/adguardhome.yaml`) with `0.0.0.0` instead of the LAN IP address of my device; this is _relatively_ safe for me because my device is not a router, but I wouldn't advise it for gateway routers as 0.0.0.0 would mean binding to the WAN port which can be unsafe.

Thank you for the advice - this did it for me. After a Docker update the container's 172.x.x.x address had changed. Setting the the `adguardhome.yaml` listen to `0.0.0.0` gives the application flexibility to listen on all interfaces, bypassing this problem. I wonder if this is something that can be integrated into the Docker installation process - there might be reasoning why the yaml captures the current IP instead of `0.0.0.0`.

**Comment by pagiraud** [comment]
I had the same problem but used a different workaround. I tried with no luck to delay the startup of Adguard by increasing the value of `START` in `/etc/init.d/adguardhome`. So, I ended up by adding in my `rc.local` : 
```
service adguardhome enable
service adguardhome start
```
It just works and seems to me way safer than putting `0.0.0.0` everywhere.
